PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/serviceexplorer/azure/appservice/ProfileFlightRecordAction.java

```diff
@@ -0,0 +1,202 @@
+            ApplicationManager.getApplication().invokeLater(() -> {
+                FlightRecorderConfiguration config = collectFlightRecorderConfiguration();
+                if (Objects.isNull(config)) {
+                    PluginUtil.showWarningNotificationProject(
+                            project,
+                            "profile flight recorder cancelled.",
+                            "Cancel profiling on app service");
+                    finishLatch.countDown();
+                    return;
+                }
+                if (config.getDuration() <= 0) {
+                    notifyUserWithErrorMessage("Invalid profile duration",
+                            "Cannot profile on app service due to "
+                                    + "configuration error: invalid "
+                                    + "duration");
+                    finishLatch.countDown();
+                    return;
+                } else {
+                    new Thread(() -> {
+                        doProfileFlightRecorder(progressIndicator,
+                                                config, finishLatch);
+                    }).start();
+                }
+            }, ModalityState.NON_MODAL);
+            finishLatch.await();
+        } catch (Exception ex) {
+            notifyUserWithErrorMessage(
+                    "Failed to profile on app service",
+                    "Cannot profile on app service, due to error: "
+                            + ex.getMessage());
+        }
+    }
+
+    private FlightRecorderConfiguration collectFlightRecorderConfiguration() {
+        RunFlightRecorderDialog ui = new RunFlightRecorderDialog(
+                project,
+                appService);
+        ui.setTitle(String.format("Start Flight Recorder on '%s'", appService.defaultHostName()));
+        ui.setOkActionListener((config) -> {
+            ui.close(DialogWrapper.OK_EXIT_CODE);
+        });
+
+        if (ui.showAndGet()) {
+            return ui.getData();
+        }
+        return null;
+    }
+
+    private void doProfileFlightRecorder(ProgressIndicator progressIndicator,
+                                         FlightRecorderConfiguration config, CountDownLatch latch) {
+        try {
+            progressIndicator.setText(String.format("start flight recorder for process(%d:%s)",
+                    config.getPid(),
+                    config.getProcessName()));
+            File file = File.createTempFile("jfr-snapshot-" + appService.name() + "-", ".jfr");
+            FlightRecorderStarterBase starter = FlightRecorderManager.getFlightRecorderStarter(appService);
+            starter.startFlightRecorder(config.getPid(), config.getDuration(), file.getName());
+            progressIndicator.setText(String.format("Recording %s seconds...", config.getDuration()));
+
+            progressIndicator.checkCanceled();
+            try {
+                Thread.sleep(ONE_SECOND * config.getDuration() + TWO_SECONDS);
```

Comment:
two seconds for buffer?

Reply:
because we need some time to send the jcmd command

Reply:
rename `TWO_SECONDS` -> `BUFFER`?

Reply:
BUFFER is confusing